[AERIE-1913]  Add simulation dataset id to `simulate` query response

skovati opened the pull request:
* **Tickets addressed:** AERIE-1913
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
<!-- What approach was taken to satisfy the ticket being addressed? What should reviewers be aware of? -->
This feature was added so that clients of the API could hit the `simulate` Hasura action, and instantly receive back the simulationDatasetId to subscribe with. This removes the need to poll the `simulate` action until it returns `status: complete`.

This was implemented by adding a new `long simulationDatasetId` field to the `ResultsProtocol` interface, and having both `PostgresResultsCellRepository` and `InMemoryResultsCellRepository` populate it. One possible issue with this implementation is that `InMemoryResultsCellRepository` doesn't really have a concept of `datasetId` so it is just getting hardcoded to 0. The justification for this is that `InMemory` is somewhat of a second class citizen only used in testing, so priority was given to the ergonomic use of `PostgresResults` instead.

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? -->
Manually tested changes in UI and GraphQL API to make sure nothing changed except returning simulationDatasetId.

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and completeness? -->
Need to update `simulate` response documentation

## Future work
<!-- What next steps can we anticipate from here, if any? -->
Further steps towards simulation result streaming
Make use of this change in the UI

skovati committed:
Add simulationDatasetId to `simulate` hasura action

skovati committed:
Add simDatasetId to `ResultsProtocol`

Implement change for InMemory and Postgres Cell Repos

skovati committed:
Use new simDatasetId in `Response`

skovati committed:
Serialize and return simDatasetId in HTTP response

camargo commented on deployment/hasura/metadata/actions.graphql:
```diff
@@ -151,6 +151,7 @@ type ResourceType {
 
 type MerlinSimulationResponse {
   status: MerlinSimulationStatus!
+  simulationDatasetId: Int
```
Can this be `Int!`?

skovati replied:
Good call! I had that in there before I was returning a datasetId in every case

skovati replied:
Change has been made!

mattdailis commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java:
```diff
@@ -290,10 +290,10 @@ private static Optional<State> getSimulationState(
-          case INCOMPLETE -> new ResultsProtocol.State.Incomplete();
-          case FAILED -> new ResultsProtocol.State.Failed(record.state().reason());
-          case SUCCESS -> new ResultsProtocol.State.Success(getSimulationResults(connection, record, planId));
+          case PENDING -> new ResultsProtocol.State.Pending(datasetId);
```
🔻 There is an unfortunate subtle difference between `simulationDatasetId` and `datasetId`. In this case, I think we want to update `SimulationDatasetRecord` to have a `simulationDatasetId` field, and pull that field out of the `record` variable here.

skovati replied:
Would this be separate from the `datasetId` field that already exists in the `SimulationDatasetRecord`?

mattdailis replied:
Correct, and this is why I say it's unfortunate 😓 

`simulation_dataset` is a join table between the `simulation` table and the `dataset` table. `datasetId` refers to the `id` field on the `dataset` table, while `simulationDatasetId` refers to the `id` field on the `simulation_dataset` table.

Being a join table, the `simulation_dataset` refers to the `dataset` table using a column of its own called `dataset_id`. So, the `SimulationDatasetRecord.datasetId` field comes from the `simulation_dataset.dataset_id` column, which is a foreign key that refers to the `id` column of the `dataset` table.

Let me know if I made things even more confusing.

skovati replied:
Ahh gotcha, so I'll add `simulation_dataset.id` to the queries in `{Get,Lookup,Create}SimulationDatasetAction` and add that into the `SimulationDatasetRecord` as `simulationDatasetId`

skovati replied:
Change has been made!

goetzrrGit reviewed (approved):
Besides Matt's suggestions, this seems pretty straightforward.

skovati committed:
Add `simulationDatasetId` to `SimulationDatasetRecord`

Additionally add new selections to existing SQL queries and populate new
record field accordingly.
